File: windows/app/ViewModels/ItemsViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;
using CommunityToolkit.Mvvm.ComponentModel;
using app.Models;
using app.Services;

namespace app.ViewModels;

public partial class ItemsViewModel : ObservableObject
{
    private readonly ApiClient _api = ApiClient.Shared;

    [ObservableProperty]
    private ObservableCollection<ItemResponse> _items = new();

    [ObservableProperty]
    private string _newName = "";

    [ObservableProperty]
    private bool _isLoading;

    [ObservableProperty]
    private bool _isSubmitting;

    [ObservableProperty]
    private string? _errorMessage;

    public async Task LoadItemsAsync(string accessToken)
    {
        IsLoading = true;
        ErrorMessage = null;
        try
        {
            var items = await _api.GetItemsAsync(accessToken);
            Items = new ObservableCollection<ItemResponse>(items);
        }
        catch (ApiException e)
        {
            ErrorMessage = e.Message;
        }
        catch (Exception e)
        {
            ErrorMessage = e.Message;
        }
        finally
        {
            IsLoading = false;
        }
    }

    public async Task AddItemAsync(string accessToken)
    {
        var name = NewName.Trim();
        if (string.IsNullOrEmpty(name)) return;

        IsSubmitting = true;
        ErrorMessage = null;
        try
        {
            var item = await _api.CreateItemAsync(accessToken, name);
            Items.Insert(0, item);
            NewName = "";
        }
        catch (ApiException e)
        {
            ErrorMessage = e.Message;
        }
        catch (Exception e)
        {
            ErrorMessage = e.Message;
        }
        finally
        {
            IsSubmitting = false;
        }
    }

    public async Task DeleteItemAsync(string accessToken, string id)
    {
        ErrorMessage = null;
        try
        {
            await _api.DeleteItemAsync(accessToken, id);
            var toRemove = Items.FirstOrDefault(i => i.Id == id);
            if (toRemove is not null) Items.Remove(toRemove);
        }
        catch (ApiException e)
        {
            ErrorMessage = e.Message;
        }
        catch (Exception e)
        {
            ErrorMessage = e.Message;
        }
    }
}

File: android/app/src/main/java/dev/mazgi/app/ItemsViewModel.kt
package dev.mazgi.app

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class ItemsUiState(
    val items: List<ItemResponse> = emptyList(),
    val isLoading: Boolean = false,
    val isSubmitting: Boolean = false,
    val newName: String = "",
    val errorMessage: String? = null,
)

class ItemsViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(ItemsUiState())
    val uiState: StateFlow<ItemsUiState> = _uiState.asStateFlow()

    private val api = APIClient.shared

    fun updateNewName(name: String) {
        _uiState.value = _uiState.value.copy(newName = name)
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null)
    }

    fun loadItems(accessToken: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val items = api.getItems(accessToken)
                _uiState.value = _uiState.value.copy(items = items)
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error")
            } finally {
                _uiState.value = _uiState.value.copy(isLoading = false)
            }
        }
    }

    fun addItem(accessToken: String) {
        val name = _uiState.value.newName.trim()
        if (name.isEmpty()) return
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSubmitting = true, errorMessage = null)
            try {
                val item = api.createItem(accessToken, name)
                _uiState.value = _uiState.value.copy(
                    items = listOf(item) + _uiState.value.items,
                    newName = "",
                )
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error")
            } finally {
                _uiState.value = _uiState.value.copy(isSubmitting = false)
            }
        }
    }

    fun deleteItem(accessToken: String, id: String) {
        viewModelScope.launch {
            try {
                api.deleteItem(accessToken, id)
                _uiState.value = _uiState.value.copy(
                    items = _uiState.value.items.filter { it.id != id },
                )
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error")
            }
        }
    }
}
